From 03e31891293a55fc23005ce32da45ac5e07e185b Mon Sep 17 00:00:00 2001
From: Pavlin Radoslavov <pavlin@google.com>
Date: Fri, 23 Feb 2018 03:35:37 -0800
Subject: [PATCH 04/59] Free the A2DP Source media alarm on the A2DP Source
 worker thread

Move the alarm_free() for the A2DP Source media alarm from
btif_a2dp_source_shutdown() to btif_a2dp_source_shutdown_delayed().
The latter is executed on the btif_a2dp_source_thread.
This avoids a race condition with btif_a2dp_source_audio_tx_stop_event()
that also might attempt to free the same alarm.

Test: Manual - unpair A2DP device while A2DP streaming.
Bug: 73741079
Change-Id: Ia871fcbeaae7c351878228e7605e65307a3d1583
(cherry picked from commit cc38c7fb4127037c3ba2c71aef0e8f32766e6785)
---
 btif/src/btif_a2dp_source.cc | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/btif/src/btif_a2dp_source.cc b/btif/src/btif_a2dp_source.cc
index b56059679..e047a5e8b 100644
--- a/btif/src/btif_a2dp_source.cc
+++ b/btif/src/btif_a2dp_source.cc
@@ -414,10 +414,6 @@ void btif_a2dp_source_shutdown(void) {
 
   APPL_TRACE_EVENT("## A2DP SOURCE STOP MEDIA THREAD ##");
 
-  // Stop the timer
-  alarm_free(btif_a2dp_source_cb.media_alarm);
-  btif_a2dp_source_cb.media_alarm = nullptr;
-
   // Exit the thread
   btif_a2dp_source_thread.DoInThread(
       FROM_HERE, base::Bind(&btif_a2dp_source_shutdown_delayed));
@@ -425,6 +421,10 @@ void btif_a2dp_source_shutdown(void) {
 }
 
 static void btif_a2dp_source_shutdown_delayed(void) {
+  // Stop the timer
+  alarm_free(btif_a2dp_source_cb.media_alarm);
+  btif_a2dp_source_cb.media_alarm = nullptr;
+
   btif_a2dp_control_cleanup();
   fixed_queue_free(btif_a2dp_source_cb.tx_audio_queue, nullptr);
   btif_a2dp_source_cb.tx_audio_queue = nullptr;
-- 
2.17.1

